### ISSUE TITLE
Speculative/MTP decoding: greedy path uses probabilistic clock-seeded sampler + 2 related concerns (dormant paths)

### ISSUE THREAD

**Issue by ohdearquant** [ohdearquant/lattice#237]
Three correctness concerns in the speculative-decoding / MTP paths, found by a hardening pass and TBV'd against the code. **All three are on dormant or opt-in paths for the shipping qwen3.5-0.8b model** (no MTP weights ship with it), so none affect default generation or the e2e-parity gate today. Filing for when MTP / self-spec model support is prioritized.

## 1. Greedy MTP path uses the probabilistic, clock-seeded rejection sampler — CONFIRMED real

`crates/inference/src/forward/metal_qwen35.rs` `generate_greedy_mtp` (~L8324):

```rust
let rs = crate::speculative::rejection_sample_draft(
    &[draft.token_id],
    std::slice::from_ref(&draft.logits),
    &verify_out.logits[0],
    std::slice::from_ref(&verify_out.logits[1]),
    false,   // <-- greedy = false
    None,    // <-- seed = None  => SpecRng::from_clock()
)
```

`rejection_sample_draft(greedy=true)` does argmax-only acceptance (`target_choice == draft_tokens[i]`) and argmax bonus — deterministic. `greedy=false` takes the probabilistic path with a **clock-seeded RNG** when `seed=None`. So a function named `generate_greedy_mtp`, entered only when `top_k<=1 && temperature<=0.0`, produces **non-deterministic** acceptance/bonus that is not argmax-equivalent to plain greedy `generate()`.

**Fix**: pass `greedy=true` (and an empty draft-logits slice, per the function's own doc) in the MTP greedy path.

**Test**: draft slot where target argmax `t != draft d`; assert `generate_greedy_mtp` output == plain greedy output token-for-token over identical logits/state.

**Reachability**: `use_mtp` requires `session.mtp.is_some()` (MTP weights loaded) + `enable_mtp`/`LATTICE_MTP` opt-in. qwen3.5-0.8b ships no MTP weights → unreachable by default.

## 2. Draft stop token short-circuits before target verification — plausible, needs confirm

`generate_greedy_mtp` (~L8289): when the draft model emits EOS/stop, the loop pushes `pending_token` and `break`s **before** verifying that draft token against the target. If the target argmax at that position is non-EOS, greedy MTP stops earlier than plain greedy would. Standard speculative decoding verifies the draft against the target and only stops if the target (or accept/bonus) is stop.

**Reachability**: same dormant MTP path as #1.

## 3. `generate_with_speculation` drops EOS from the output stream — plausible, test-only helper

`crates/inference/src/speculative.rs` `generate_with_speculation` (~L1920/1932/1946) returns `generated` without appending EOS in several branches, whereas `generate.rs` pushes EOS into `generated_ids` before breaking (L291-310). The helper has **no non-test callers** today; its tests (~L2452/2455/2539) assert the current (no-EOS) behavior, so any change must update the contract intentionally — verify whether the omission is by-design for the helper's intended use before "fixing".

---

Notes: #1 is the only one verified to be a genuine defect (via `rejection_sample_draft` semantics). #2/#3 are plausible from code reading and need a differential test to confirm. #1 and #2 live in `metal_qwen35.rs`; fixes there are deferred until #235 (chunked GDN prefill) merges to avoid file-contention on that large file.

**Comment by ohdearquant** [comment]
## Concern #1 — FIXED (merged)

Fixed in #286, merged to `main` as `e9b19c20a`. The greedy MTP call site (`generate_greedy_mtp`, `metal_qwen35.rs`) now passes `greedy=true` to `rejection_sample_draft`, giving deterministic argmax-only acceptance/bonus that is token-for-token equivalent to plain greedy `generate()`. A new unit test (`rejection_greedy_one_draft_mtp_contract_empty_draft_logits`) locks the n=1 greedy contract; the sibling `mtp_verify_draft` (`speculative.rs:1284`) was verified to be the intended probabilistic ADR-050 verifier and left unchanged. Full CI green incl. the `--features f16,metal-gpu` clippy gate and e2e-parity.

## Concern #2 — CONFIRMED real (still open)

Verified by reading `generate_greedy_mtp` at `metal_qwen35.rs:8749-8753` (current `main`):

```rust
if is_stop(draft.token_id) {
    // Draft is EOS — flush pending, stop before draft.
    generated_ids.push(pending_token);
    break;
}
```

When the **draft** model emits a stop token, the loop flushes `pending_token` and `break`s **before running the verify phase**. The draft is the cheap proposal model; its EOS guess is never checked against the target. The target's own prediction for the position after `pending_token` is `verify_out.logits[0]` (argmax) — if that is non-stop, plain greedy `generate()` would continue, but greedy MTP terminates. So the cheap draft can truncate generation early, diverging from plain greedy.

The fix is **not** a one-line flag flip (unlike #1). It needs to fold the draft-EOS case into the verify path with correct stop edge-case handling:
- draft EOS but target argmax non-stop → reject, emit `pending`, continue with `pending = argmax(verify_out.logits[0])`
- draft EOS and target argmax also EOS → emit `pending` (+ EOS per `generate.rs` EOS-in-output convention), then stop — must **not** generate a bonus token past EOS (the current accept branch at `8819` would set `pending = bonus` and loop)

C(τ)≈0.5. Same dormant path as #1 (qwen3.5-0.8b ships no MTP weights, so e2e-parity does not exercise it); the control-flow fix can be unit-tested at the verifier-semantics level but the Metal wiring is review-gated. Worth a dedicated PR with a differential test against plain greedy on synthetic logits.

## Concern #3 — still open, likely by-design

`generate_with_speculation` (`speculative.rs`) returning `generated` without appending EOS: it has no non-test callers, and its existing tests assert the no-EOS behavior. This is plausibly intentional for the helper's contract rather than a defect. Needs a decision on the intended contract before any change — not obviously a bug.

---

Leaving this issue open for #2 and #3.

**Comment by ohdearquant** [comment]
> _Status update by Claude (Anthropic agent) on behalf of @ohdearquant._

Triage during the hardening loop — two of the three concerns are now resolved:

- **Concern #1 (clock-seeded greedy MTP sampler): FIXED** by #286 (commit `e9b19c20a`). `generate_greedy_mtp` now passes `greedy=true` to `rejection_sample_draft` (`metal_qwen35.rs:8813`), so the verifier is deterministic argmax-only — no more `from_clock()` RNG. Code comment at L8798-8803 documents the contract.
- **Concern #2 (draft-stop short-circuit before target verify): FIXED** by #287. `generate_greedy_mtp` now verifies the draft-EOS against the target before terminating.
- **Concern #3 (`generate_with_speculation` drops EOS): OPEN, low-value.** This is the test-only helper with no non-test callers; its tests assert the current no-EOS behavior, and the original report flagged the omission as possibly by-design. Leaving open but de-scoped — recommend either an intentional contract decision or wontfix rather than a reflexive "fix".

The sibling self-spec path (`generate_greedy_self_spec`, concern equivalent to #2 + a prefill-EOS Case-A bug) is tracked separately in #289 and is the active follow-up.

**Comment by ohdearquant** [comment]
> _Comment by Claude (Anthropic agent) on behalf of @ohdearquant._

## Resolved on main — closing. All three concerns dispositioned with code evidence.

A TBV re-read against current `main` (5cb464b5b) shows #1 and #2 were both fixed by **#286** (`fix(inference): greedy MTP path must use deterministic argmax verifier`, commit e9b19c20a, ancestor of HEAD), and #3 is a by-design test-only helper, not a defect.

**Concern #1 — greedy MTP uses clock-seeded probabilistic sampler — FIXED.**
`generate_greedy_mtp` now calls `rejection_sample_draft(..., greedy=true, seed=None)` (`metal_qwen35.rs:8808-8815`). `greedy=true` takes the deterministic argmax-acceptance path; the `seed` is unread in greedy mode. The call site carries a comment documenting exactly this regression and its fix (L8793-8807).

**Concern #2 — draft stop token short-circuits before target verification — FIXED.**
The draft phase no longer breaks on a draft EOS before verification. `metal_qwen35.rs:8767-8772`:
> `// NOTE: We intentionally do NOT short-circuit on draft EOS here. The draft's EOS guess must be verified against the target model. ... Short-circuiting before verify caused concern #2 of #237: a wrong draft-EOS would silently truncate generation early.`
The draft EOS is now carried into `verify_tokens_batched`/`mtp_greedy_round` and only stops generation if the target agrees.

**Concern #3 — `generate_with_speculation` omits EOS — by design, not a defect.**
This helper has no non-test callers, and its tests assert the current (no-EOS) behavior. As the issue itself flagged, "any change must update the contract intentionally." It is not a live correctness bug; if the helper ever gains a production caller, the EOS contract should be revisited at that point and tracked under a fresh issue.

Both live concerns are resolved on main and reachable only via the dormant MTP opt-in (no MTP weights ship with qwen3.5-0.8b), so default generation and the e2e-parity gate are unaffected either way. Closing as resolved.

_Verified via an autonomous hardening loop: fix commit confirmed an ancestor of HEAD; both call sites read line-by-line against source._
